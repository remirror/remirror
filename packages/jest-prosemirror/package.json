{
  "name": "jest-prosemirror",
  "version": "3.0.0-beta.4",
  "description": "Write expressive tests for your prosemirror editor",
  "keywords": [
    "jest",
    "prosemirror",
    "snapshots",
    "tests"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-prosemirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-prosemirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-prosemirror.js",
      "require": "./dist/jest-prosemirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-prosemirror-environment.js",
      "require": "./environment/dist/jest-prosemirror-environment.cjs"
    },
    "./package.json": "./package.json",
    "./serializer": {
      "types": "./dist-types/serializer.d.ts",
      "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
    }
  },
  "main": "./dist/jest-prosemirror.cjs",
  "module": "./dist/jest-prosemirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment",
    "serializer"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core-constants": "3.0.0-beta.3",
    "@remirror/core-helpers": "4.0.0-beta.3",
    "@remirror/core-types": "3.0.0-beta.4",
    "@remirror/core-utils": "3.0.0-beta.4",
    "@remirror/pm": "3.0.0-beta.4",
    "@testing-library/dom": "^9.3.0",
    "@types/stringify-object": "^3.3.1",
    "chalk": "^4.1.0",
    "jest-matcher-utils": "^29.5.0",
    "prosemirror-schema-basic": "^1.2.3",
    "prosemirror-tables": "^1.3.7",
    "prosemirror-test-builder": "^1.1.1",
    "stringify-object": "^3.3.0",
    "test-keyboard": "2.0.7-beta.4"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-prosemirror.d.ts",
        "import": "./dist/jest-prosemirror.js",
        "require": "./dist/jest-prosemirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-prosemirror-environment.d.ts",
        "import": "./environment/dist/jest-prosemirror-environment.js",
        "require": "./environment/dist/jest-prosemirror-environment.cjs"
      },
      "./package.json": "./package.json",
      "./serializer": {
        "types": "./serializer/dist/jest-prosemirror-serializer.d.ts",
        "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
      }
    },
    "main": "./dist/jest-prosemirror.cjs",
    "module": "./dist/jest-prosemirror.js",
    "type": "module",
    "types": "./dist/jest-prosemirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts",
      "serializer.cts"
    ]
  }
}
