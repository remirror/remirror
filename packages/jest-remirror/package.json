{
  "name": "jest-remirror",
  "version": "3.0.0-beta.6",
  "description": "A text editor for react built with prosemirror",
  "keywords": [
    "prosemirror",
    "react",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-remirror.js",
      "require": "./dist/jest-remirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-remirror-environment.js",
      "require": "./environment/dist/jest-remirror-environment.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/jest-remirror.cjs",
  "module": "./dist/jest-remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.6",
    "@remirror/pm": "3.0.0-beta.4",
    "@remirror/preset-core": "3.0.0-beta.6",
    "@testing-library/dom": "^9.3.0",
    "@types/sanitize-html": "^2.9.0",
    "jest-prosemirror": "3.0.0-beta.4",
    "sanitize-html": "^2.10.0",
    "tiny-warning": "^1.0.3"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "@types/node": "^16.18.34",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "@types/node": "*",
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    },
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-remirror.d.ts",
        "import": "./dist/jest-remirror.js",
        "require": "./dist/jest-remirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-remirror-environment.d.ts",
        "import": "./environment/dist/jest-remirror-environment.js",
        "require": "./environment/dist/jest-remirror-environment.cjs"
      },
      "./package.json": "./package.json"
    },
    "main": "./dist/jest-remirror.cjs",
    "module": "./dist/jest-remirror.js",
    "type": "module",
    "types": "./dist/jest-remirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts"
    ]
  }
}
