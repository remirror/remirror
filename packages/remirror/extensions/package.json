{
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/remirror-extensions.d.ts",
      "import": "./dist/remirror-extensions.js",
      "require": "./dist/remirror-extensions.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/remirror-extensions.cjs",
  "module": "./dist/remirror-extensions.js",
  "types": "./dist/remirror-extensions.d.ts"
}
