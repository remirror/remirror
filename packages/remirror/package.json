{
  "name": "remirror",
  "version": "3.0.0-beta.8",
  "description": "One editing package to rule them all, one editing package to bind them.",
  "keywords": [
    "prosemirror",
    "react",
    "remirror",
    "wysiwyg",
    "editing",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>",
    "ocavue "
  ],
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/remirror.js",
      "require": "./dist/remirror.cjs"
    },
    "./dom": {
      "types": "./dist-types/dom.d.ts",
      "import": "./dom/dist/remirror-dom.js",
      "require": "./dom/dist/remirror-dom.cjs"
    },
    "./extensions": {
      "types": "./dist-types/extensions.d.ts",
      "import": "./extensions/dist/remirror-extensions.js",
      "require": "./extensions/dist/remirror-extensions.cjs"
    },
    "./icons": {
      "types": "./dist-types/icons.d.ts",
      "import": "./icons/dist/remirror-icons.js",
      "require": "./icons/dist/remirror-icons.cjs"
    },
    "./package.json": "./package.json",
    "./styles/*": "./styles/*"
  },
  "main": "./dist/remirror.cjs",
  "module": "./dist/remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "dom",
    "extensions",
    "icons",
    "styles"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.8",
    "@remirror/core-constants": "3.0.0-beta.4",
    "@remirror/core-helpers": "4.0.0-beta.5",
    "@remirror/core-types": "3.0.0-beta.5",
    "@remirror/core-utils": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.8",
    "@remirror/extension-annotation": "3.0.0-beta.8",
    "@remirror/extension-bidi": "3.0.0-beta.8",
    "@remirror/extension-blockquote": "3.0.0-beta.8",
    "@remirror/extension-bold": "3.0.0-beta.8",
    "@remirror/extension-callout": "3.0.0-beta.8",
    "@remirror/extension-code": "3.0.0-beta.8",
    "@remirror/extension-code-block": "3.0.0-beta.8",
    "@remirror/extension-collaboration": "3.0.0-beta.8",
    "@remirror/extension-columns": "3.0.0-beta.8",
    "@remirror/extension-diff": "3.0.0-beta.8",
    "@remirror/extension-doc": "3.0.0-beta.8",
    "@remirror/extension-drop-cursor": "3.0.0-beta.8",
    "@remirror/extension-embed": "3.0.0-beta.8",
    "@remirror/extension-emoji": "3.0.0-beta.8",
    "@remirror/extension-entity-reference": "3.0.0-beta.8",
    "@remirror/extension-epic-mode": "3.0.0-beta.8",
    "@remirror/extension-events": "3.0.0-beta.8",
    "@remirror/extension-find": "1.0.0-beta.8",
    "@remirror/extension-font-family": "3.0.0-beta.8",
    "@remirror/extension-font-size": "3.0.0-beta.8",
    "@remirror/extension-gap-cursor": "3.0.0-beta.8",
    "@remirror/extension-hard-break": "3.0.0-beta.8",
    "@remirror/extension-heading": "3.0.0-beta.8",
    "@remirror/extension-history": "3.0.0-beta.8",
    "@remirror/extension-horizontal-rule": "3.0.0-beta.8",
    "@remirror/extension-image": "3.0.0-beta.8",
    "@remirror/extension-italic": "3.0.0-beta.8",
    "@remirror/extension-link": "3.0.0-beta.8",
    "@remirror/extension-list": "3.0.0-beta.8",
    "@remirror/extension-markdown": "3.0.0-beta.8",
    "@remirror/extension-mention": "3.0.0-beta.8",
    "@remirror/extension-mention-atom": "3.0.0-beta.8",
    "@remirror/extension-node-formatting": "3.0.0-beta.8",
    "@remirror/extension-paragraph": "3.0.0-beta.8",
    "@remirror/extension-placeholder": "3.0.0-beta.8",
    "@remirror/extension-positioner": "3.0.0-beta.8",
    "@remirror/extension-shortcuts": "3.0.0-beta.8",
    "@remirror/extension-strike": "3.0.0-beta.8",
    "@remirror/extension-sub": "3.0.0-beta.8",
    "@remirror/extension-sup": "3.0.0-beta.8",
    "@remirror/extension-tables": "3.0.0-beta.8",
    "@remirror/extension-text": "3.0.0-beta.8",
    "@remirror/extension-text-case": "3.0.0-beta.8",
    "@remirror/extension-text-color": "3.0.0-beta.8",
    "@remirror/extension-text-highlight": "3.0.0-beta.8",
    "@remirror/extension-trailing-node": "3.0.0-beta.8",
    "@remirror/extension-underline": "3.0.0-beta.8",
    "@remirror/extension-whitespace": "3.0.0-beta.8",
    "@remirror/icons": "3.0.0-beta.5",
    "@remirror/preset-core": "3.0.0-beta.8",
    "@remirror/preset-formatting": "3.0.0-beta.8",
    "@remirror/preset-wysiwyg": "3.0.0-beta.8",
    "@remirror/theme": "3.0.0-beta.5",
    "@types/refractor": "^3.0.2",
    "refractor": "^3.6.0"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.1",
    "@remirror/pm": "3.0.0-beta.6",
    "@types/prettier": "^2.7.2",
    "prettier": "^2.8.8"
  },
  "peerDependencies": {
    "@remirror/pm": "^3.0.0-beta.6",
    "@types/prettier": "^2.7.2",
    "prettier": "^2.8.8"
  },
  "peerDependenciesMeta": {
    "@types/prettier": {
      "optional": true
    },
    "prettier": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/remirror.d.ts",
        "import": "./dist/remirror.js",
        "require": "./dist/remirror.cjs"
      },
      "./dom": {
        "types": "./dom/dist/remirror-dom.d.ts",
        "import": "./dom/dist/remirror-dom.js",
        "require": "./dom/dist/remirror-dom.cjs"
      },
      "./extensions": {
        "types": "./extensions/dist/remirror-extensions.d.ts",
        "import": "./extensions/dist/remirror-extensions.js",
        "require": "./extensions/dist/remirror-extensions.cjs"
      },
      "./icons": {
        "types": "./icons/dist/remirror-icons.d.ts",
        "import": "./icons/dist/remirror-icons.js",
        "require": "./icons/dist/remirror-icons.cjs"
      },
      "./package.json": "./package.json",
      "./styles/*": "./styles/*"
    },
    "main": "./dist/remirror.cjs",
    "module": "./dist/remirror.js",
    "type": "module",
    "types": "./dist/remirror.d.ts"
  },
  "@remirror": {},
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "dom.ts",
      "extensions.ts",
      "icons.ts"
    ]
  }
}
