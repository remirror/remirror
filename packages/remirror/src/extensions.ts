// Extensions
export * from '@remirror/extension-annotation';
export * from '@remirror/extension-bidi';
export * from '@remirror/extension-blockquote';
export * from '@remirror/extension-bold';
export * from '@remirror/extension-callout';
export * from '@remirror/extension-code';
export * from '@remirror/extension-code-block';
export * from '@remirror/extension-codemirror5';
export * from '@remirror/extension-collaboration';
export * from '@remirror/extension-columns';
export * from '@remirror/extension-diff';
export * from '@remirror/extension-doc';
export * from '@remirror/extension-drop-cursor';
export * from '@remirror/extension-embed';
export * from '@remirror/extension-emoji';
export * from '@remirror/extension-entity-reference';
export * from '@remirror/extension-epic-mode';
export * from '@remirror/extension-events';
export * from '@remirror/extension-find';
export * from '@remirror/extension-font-family';
export * from '@remirror/extension-font-size';
export * from '@remirror/extension-gap-cursor';
export * from '@remirror/extension-hard-break';
export * from '@remirror/extension-heading';
export * from '@remirror/extension-history';
export * from '@remirror/extension-horizontal-rule';
export * from '@remirror/extension-image';
export * from '@remirror/extension-italic';
export * from '@remirror/extension-link';
export * from '@remirror/extension-list';
export * from '@remirror/extension-markdown';
export * from '@remirror/extension-mention';
export * from '@remirror/extension-mention-atom';
export * from '@remirror/extension-node-formatting';
export * from '@remirror/extension-paragraph';
export * from '@remirror/extension-placeholder';
export * from '@remirror/extension-positioner';
export * from '@remirror/extension-shortcuts';
export * from '@remirror/extension-strike';
export * from '@remirror/extension-sub';
export * from '@remirror/extension-sup';
export * from '@remirror/extension-tables';
export * from '@remirror/extension-text';
export * from '@remirror/extension-text-case';
export * from '@remirror/extension-text-color';
export * from '@remirror/extension-text-highlight';
export * from '@remirror/extension-trailing-node';
export * from '@remirror/extension-underline';
export * from '@remirror/extension-whitespace';
export * from '@remirror/extension-yjs';

// Presets
export * from '@remirror/preset-core';
export * from '@remirror/preset-formatting';
export * from '@remirror/preset-wysiwyg';
