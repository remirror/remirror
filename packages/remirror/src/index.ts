export * from '@remirror/core';
export * from '@remirror/theme';
