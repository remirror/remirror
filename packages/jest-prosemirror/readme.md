# jest-prosemirror

[![npm](https://img.shields.io/npm/dm/jest-prosemirror.svg?&logo=npm)](https://www.npmjs.com/package/jest-prosemirror)

## The problem

You want to write tests for some of your prosemirror editor but you don't know where to start. You know you should avoid testing implementation details and just want to be sure that your commands and plugins produce the correct underlying prosemirror state.

## This solution

`jest-prosemirror` takes inspiration from the [`testing-library`](https://github.com/testing-library/react-testing-library) mantra and enables you to write more intuitive tests for your prosemirror editor.

## Installation

```bash
yarn add jest-prosemirror # yarn
pnpm add jest-prosemirror # pnpm
npm install jest-prosemirror # npm
```

## Getting started

### Quick setup

For a quick setup add the following to your jest.config.js file.

```js
module.exports = {
  setupFilesAfterEnv: ['jest-prosemirror/environment'],
  testEnvironment: 'jsdom', // Required for dom manipulation
};
```

This will automatically:

- Add the jest assertions `toTransformNode` and `toEqualProsemirrorNode`.

If you are using typescript then add this to your `tsconfig.json` file for global type support.

```json
{
  "compilerOptions": {
    "types": ["jest-prosemirror"]
  }
}
```

### Manual setup

Create a `jest.framework.dom.ts` file and add the following

```ts
// jest.framework.dom.ts

import { prosemirrorMatchers } from 'jest-prosemirror';

// Add jest-prosemirror assertions
expect.extend(prosemirrorMatchers);
```

In your `jest.config.js` file add this to the configuration

```js
module.exports = {
  setupFilesAfterEnv: ['<rootDir>/jest.framework.dom.ts'],
  testEnvironment: 'jsdom', // Required for dom manipulation
};
```

## Snapshot serializer

This package exports a serializer for better snapshot testing of prosemirror primitives. To set this up add the following to your `jest.config.js` file.

```js
module.exports = {
  snapshotSerializers: ['jest-prosemirror/serializer'],
};
```

Alternatively, you can add the following to your `jest.framework.dom.ts` file.

```ts
// jest.framework.dom.ts

import { prosemirrorSerializer } from 'jest-prosemirror';

// Add the serializer for use throughout all the configured test files.
expect.addSnapshotSerializer(prosemirrorSerializer);
```

## Jest Matchers

### `toTransformNode`

A utility from jest-prosemirror which tests that a command transforms the prosemirror node in the desired way.

```ts
import { doc, p, schema, strong } from 'jest-prosemirror';
import { removeMark } from 'remirror/core/utils';

test('remove the mark', () => {
  const type = schema.marks.bold;
  const from = doc(p(strong('<start>bold<end>')));
  const to = doc(p('bold'));

  expect(removeMark({ type })).toTransformNode({ from, to });
});
```

This tests that mark has been removed by the provided command.

The `to` property is optional and can be left blank to test that the node is identical after the transform.

### `toEqualProsemirrorNode`

Tests that two prosemirror documents are equal. Pass in the expected document and it checks that they are the same.

```ts
import { createEditor, doc, p } from 'jest-prosemirror';
import { removeNodeAtPosition } from 'remirror/core/utils';

test('remove block top level node at specified position', () => {
  const {
    state: { tr },
  } = createEditor(doc(p('x'), p('one')));
  const newTr = removeNodeAtPosition({ pos: 3, tr });

  expect(newTr).not.toBe(tr);
  expect(newTr.doc).toEqualProsemirrorNode(doc(p('x')));
});
```

## `createEditor`

Create a test prosemirror editor.

The call to create editor can be chained with various commands to enable testing of the editor at each step without the need for intermediate holding variables.

```ts
import { createEditor, doc, p } from 'jest-remirror';
import { suggest } from 'prosemirror-suggest';

test('`keyBindings`', () => {
  const keyBindings = {
    Enter: jest.fn((params: SuggestKeyBindingParameter) => {
      params.command();
    }),
  };

  const plugin = suggest({
    char: '@',
    name: 'at',
    keyBindings,
    matchOffset: 0,
    createCommand:
      ({ view }) =>
      () =>
        view.dispatch(view.state.tr.insertText('awesome')),
  });

  createEditor(doc(p('<cursor>')), { plugins: [plugin] })
    .insertText('@')
    .press('Enter')
    .callback((content) => {
      expect(content.state.doc).toEqualProsemirrorNode(doc(p('@awesome')));
    });
});
```

### Parameters

#### taggedDoc - `TaggedProsemirrorNode`

The tagged prosemirror node to inject into the editor.

#### options - `CreateEditorOptions`

The following options which include all [DirectEditorProps](http://prosemirror.net/docs/ref/#view.DirectEditorProps) except for `state`.

| **Property** | **Type** | **Default** | **Description** |
| --- | :-: | :-: | --- |
| `autoClean` | `boolean` | `true` | Whether to auto remove the editor from the dom after each test. It is advisable to leave this unchanged. |
| `plugins` | `Plugin[]` | `[]` | The plugins that the test editor should use. |
| `rules` | `InputRule[]` | `[]` | The input rules that the test editor should use. |

### Returns

#### start - `number`

The start of the current selection.

#### end - `number | undefined`

The end of the current selection.

#### selection - `Selection`

The current prosemirror selection

#### schema - `EditorSchema`

The prosemirror schema.

#### state - `EditorState`

The prosemirror state.

#### view - `EditorView`

The prosemirror view.

#### overwrite - `(doc: TaggedProsemirrorNode) => ReturnType<typeof createEditor>`

Overwrite all the current content within the editor.

**`doc`** - the new content to use.

#### command - `(command: CommandFunction) => ReturnType<typeof createEditor>`

Run the command within the prosemirror editor.

```ts
test('commands are run', () => {
  createEditor(doc(p('<cursor>')))
    .command((state, dispatch) => {
      if (dispatch) {
        dispatch(state.tr.insertText('hello'));
      }
    })
    .callback((content) => {
      expect(content.state.doc).toEqualProsemirrorDocument(doc(p('hello')));
    });
});
```

**`command`** - the command function to run.

#### insertText - `(text: string) => ReturnType<typeof createEditor>`

Insert text into the editor at the current position.

**`text`** - the text to insert

#### jumpTo - `(start: 'start' | 'end' | number, end?: number) => ReturnType<typeof createEditor>`

Jump to the specified position in the editor.

**`start`** - a number position or the shorthand 'start' | 'end'. **`[end]`** - the option end position of the new selection.

#### shortcut - `(mod: string) => ReturnType<typeof createEditor>`

Type a keyboard shortcut - e.g. `Mod-Enter`.

**NOTE** This only simulates the events. For example an `Mod-Enter` would run all enter key handlers but not actually create a new line. I'd welcome a pull request to fix this shortcoming.

**`mod`** - the keyboard shortcut to type

#### press - `(char: string) => ReturnType<typeof createEditor>`

Simulate a keypress which is run through the editor's key handlers.

**NOTE** This only simulates the events. For example an `Enter` would run all enter key handlers but not actually create a new line. I'd welcome a pull request to fix this shortcoming.

**`char`** - the character to type

#### fire - `(params: Omit<FireEventAtPositionParameter, 'view'>) => ReturnType<typeof createEditor>`

Fire an event in the editor. This api is not the most well tested and can be quite flakey.

**`params`** - the fire event parameters

#### callback - `(fn: (content: ReturnValueCallbackParameter) => void) => ReturnType<typeof createEditor>`

Callback function which receives the `start`, `end`, `state`, `view`, `schema` and `selection` properties and allows for easier testing of the current state of the editor.

## Acknowledgements

This package borrows very heavily from [prosemirror-test-builder](https://www.npmjs.com/package/prosemirror-test-builder)
