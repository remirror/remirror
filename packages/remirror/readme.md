# remirror

> One package to rule them all, one entry point to bind them.

[![Version][version]][npm] [![Weekly Downloads][downloads-badge]][npm] [![Bundled size][size-badge]][size] [![Typed Codebase][typescript]](#) [![MIT License][license]](#)

[version]: https://flat.badgen.net/npm/v/remirror
[npm]: https://npmjs.com/package/remirror
[license]: https://flat.badgen.net/badge/license/MIT/purple
[size]: https://bundlephobia.com/result?p=remirror
[size-badge]: https://flat.badgen.net/bundlephobia/minzip/remirror
[typescript]: https://flat.badgen.net/badge/icon/TypeScript?icon=typescript&label
[downloads-badge]: https://badgen.net/npm/dw/remirror/red?icon=npm

## Installation

```bash
# yarn
yarn add remirror @remirror/pm

# pnpm
pnpm add remirror @remirror/pm

# npm
npm install remirror @remirror/pm
```

The `remirror` package will automatically install the `@remirror/pm` package for you. You only need to install it yourself if you'd like to ensure consistent versions of the `prosemirror-*` libraries when importing from `@remirror/pm/state` instead of `prosemirror-state` or `@remirror/pm/model` instead of `prosemirror-model`.

## Usage

Rather than installing multiple scoped packages, the `remirror` package is a gateway to using all the goodness that remirror provides while minimising your bundle size.

The following creates a controlled editor with React.

```tsx
import React from 'react';
import { SocialPreset } from 'remirror/extensions';
import { Remirror, SocialEmojiComponent, useRemirror } from '@remirror/react';

const EditorWrapper = () => {
  const socialPreset = new SocialPreset();
  const { state, onChange } = useRemirror({ extensions: () => [...socialPreset()] });

  return (
    <Remirror state={state} onChange={onChange} manager={manager} autoRender={true}>
      <SocialEmojiComponent />
    </Remirror>
  );
};
```

These are the entry points available through the `remirror` package.

- `remirror` - All the core functionality available through `@remirror/core`.
- `remirror/extensions` - All the core extensions and presets made available through the main `remirror` repository. This doesn't include any framework specific extensions and presets.
- `remirror/dom` - The dom framework implementation of via `createDomEditor`.
