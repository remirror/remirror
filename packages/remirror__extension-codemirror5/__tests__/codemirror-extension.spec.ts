import 'codemirror/mode/javascript/javascript';
import 'codemirror/mode/meta';
import 'codemirror/mode/python/python';
import 'codemirror/mode/yaml/yaml';

import CodeMirror from 'codemirror';
import { pmBuild } from 'jest-prosemirror';
import { extensionValidityTest, renderEditor } from 'jest-remirror';
import { htmlToProsemirrorNode, object, prosemirrorNodeToHtml } from 'remirror';
import { createCoreManager } from 'remirror/extensions';

import { CodeMirrorExtension, CodeMirrorExtensionOptions } from '../';

extensionValidityTest(CodeMirrorExtension);

function getCodeMirrorInstance(dom: Element): CodeMirror.Editor | undefined {
  return (dom.querySelector('.CodeMirror') as any).CodeMirror;
}

function getCodeMirrorMode(dom: Element): string | null | undefined {
  return getCodeMirrorInstance(dom)?.getMode()?.name;
}

describe('schema', () => {
  const { schema } = createCoreManager([new CodeMirrorExtension({ CodeMirror })]);
  const content = 'hello world!';
  const { codeMirror, doc } = pmBuild(schema, {
    codeMirror: { nodeType: 'codeMirror' },
  });

  it('creates the correct dom node', () => {
    expect(prosemirrorNodeToHtml(codeMirror(content))).toBe(`<pre><code>${content}</code></pre>`);
  });

  it('parses the dom structure and finds itself', () => {
    const node = htmlToProsemirrorNode({
      schema,
      content: `<pre><code>${content}</code></pre>`,
    });
    const expected = doc(codeMirror(content));

    expect(node).toEqualProsemirrorNode(expected);
  });
});

function create(options: CodeMirrorExtensionOptions = object()) {
  const renderEditorReturn = renderEditor([new CodeMirrorExtension({ ...options, CodeMirror })]);

  const {
    add,
    nodes: { doc, p },
    commands,
    view,
    attributeNodes: { codeMirror },
  } = renderEditorReturn;
  const { dom } = view;
  const plainBlock = codeMirror({ codeMirrorConfig: { mode: 'text/plain' } });
  const pyBlock = codeMirror({ codeMirrorConfig: { mode: 'text/x-python' } });
  return {
    add,
    view,
    doc,
    p,
    codeMirror,
    plainBlock,
    pyBlock,
    dom,
    commands,
  };
}

describe('nodeView', () => {
  it('renders the CodeMirror editor', () => {
    const { add, doc, plainBlock, dom } = create();

    add(doc(plainBlock(`hello`)));
    expect(dom.querySelector('.CodeMirror')).not.toBeNil();
  });

  describe('syntax highlight', () => {
    it('reanders Python correctly', () => {
      const { add, doc, pyBlock, dom } = create();
      // In Python, `print` is a builtin function.
      add(doc(pyBlock(`print('hi')`)));
      expect(dom.querySelector('.cm-builtin')!.innerHTML).toMatch('print');
    });

    it('reanders plain text correctly', () => {
      const { add, doc, plainBlock, dom } = create();
      add(doc(plainBlock(`print('hi')`)));
      expect(dom.querySelector('.cm-builtin')).toBeNil();
    });
  });

  describe('load mode by language name', () => {
    for (const [inputLanguage, expectedMode] of [
      ['javascript', 'javascript'],
      ['JavaScript', 'javascript'],
      ['JS', 'javascript'],
      ['Python', 'python'],
      ['yaml', 'yaml'],
      ['yml', 'yaml'],
    ]) {
      it(`can parse "${inputLanguage}"`, () => {
        const { add, doc, codeMirror, dom } = create();

        const codeMirrorNode = codeMirror({ language: inputLanguage })('');
        add(doc(codeMirrorNode));
        expect(getCodeMirrorMode(dom)).toEqual(expectedMode);
      });
    }
  });

  it('can be updated', () => {
    const { add, plainBlock, doc, dom } = create();
    add(doc(plainBlock(`hello`)))
      .callback(() => {
        expect(dom.querySelector('.CodeMirror-line')!.innerHTML).toMatchSnapshot();
      })
      .insertText(' world')
      .callback(() => {
        expect(dom.querySelector('.CodeMirror-line')!.innerHTML).toMatchSnapshot();
      });
  });

  describe('selection', () => {
    it('can be selected around', () => {
      const { add, plainBlock, doc, p, view } = create();
      const chain = add(doc(p('1<start>2'), plainBlock(`3`), p('4<end>5')));
      expect(chain.from).toBe(2);
      expect(chain.from).toBe(view.state.selection.from);
      expect(chain.to).toBe(9);
      expect(chain.to).toBe(view.state.selection.to);
    });

    it('can be selected around and deleted', () => {
      const { add, plainBlock, doc, p } = create();

      for (const key of ['Delete', 'Backspace']) {
        const chain = add(doc(p('1<start>2'), plainBlock(`3`), p('4<end>5'))).press(key);
        expect(chain.from).toBe(2);
        expect(chain.to).toBe(2);
        expect(chain.doc).toEqualRemirrorDocument(doc(p('15')));
      }
    });

    it('can set selection inside the codemirror', async () => {
      const { add, plainBlock, doc, p } = create();

      const chain = add(doc(p('12'), plainBlock(`34567`), p('89')))
        .selectText({ from: 6, to: 9 })
        .backspace();
      expect(chain.doc).toEqualRemirrorDocument(doc(p('12'), plainBlock('37'), p('89')));
    });
  });

  // JSDom has a bad implement for keyboard events. We will need to do lots of mock
  // in order to test the keyboard interaction between CodeMirror and Prosemirror.
  // jest-playwright might be a better solution for this.
  it.todo('keymap');
});

describe('commands', () => {
  describe('createCodeMirror', () => {
    const { view, add, codeMirror, doc, p, commands } = create();
    const jsBlock = codeMirror({ language: 'javascript' });

    it('creates the codeBlock', () => {
      add(doc(p(`<cursor>`)));
      commands.createCodeMirror({ language: 'javascript' });
      expect(view.state.doc).toEqualRemirrorDocument(doc(jsBlock('')));
    });
  });

  describe('updateCodeMirror', () => {
    const { add, codeMirror, doc, dom } = create();
    const jsBlock = codeMirror({ language: 'javascript' });

    it('updates the language', () => {
      add(doc(jsBlock())).callback((content) => {
        expect(getCodeMirrorMode(dom)).toBe('javascript');

        content.commands.updateCodeMirror({ language: 'python' });
        expect(getCodeMirrorMode(dom)).toBe('python');
      });
    });
  });
});
