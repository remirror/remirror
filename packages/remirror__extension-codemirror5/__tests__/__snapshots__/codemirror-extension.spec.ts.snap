// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodeView can be updated 1`] = `
<span
  style="padding-right: .1px;"
  role="presentation"
>
  hello
</span>
`;

exports[`nodeView can be updated 2`] = `
<span
  style="padding-right: .1px;"
  role="presentation"
>
  hello world
</span>
`;
